File: gdscript/src/main/java/gdscript/psi/impl/GdGetMethodIdRefImpl.java
package gdscript.psi.impl;

import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElementVisitor;
import gdscript.psi.GdGetMethodIdRef;
import gdscript.psi.GdVisitor;
import org.jetbrains.annotations.NotNull;

public class GdGetMethodIdRefImpl extends GdRefElementImpl implements GdGetMethodIdRef {

  public GdGetMethodIdRefImpl(ASTNode node) {
    super(node);
  }

  public void accept(@NotNull GdVisitor visitor) {
    visitor.visitGetMethodIdRef(this);
  }

  @Override
  public void accept(@NotNull PsiElementVisitor visitor) {
    if (visitor instanceof GdVisitor) accept((GdVisitor)visitor);
    else super.accept(visitor);
  }

}

File: gdscript/src/main/kotlin/gdscript/psi/impl/GdGetMethodIdRefImpl.kt
package gdscript.psi.impl

import com.intellij.lang.ASTNode
import com.intellij.model.psi.PsiSymbolReference
import com.intellij.polySymbols.references.polySymbolOwnReferences
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.util.PsiTreeUtil
import gdscript.polySymbols.GdPolySymbolKind.METHOD
import gdscript.polySymbols.GdPolySymbolModifier.STATIC
import gdscript.polySymbols.scope.hasModifier
import gdscript.psi.GdClassVarDeclTl
import gdscript.psi.GdGetMethodIdRef
import gdscript.psi.GdVisitor
import org.jetbrains.annotations.Unmodifiable

class GdGetMethodIdRefImpl(node: ASTNode) : GdRefElementImpl(node), GdGetMethodIdRef {
    fun accept(visitor: GdVisitor) {
        visitor.visitGetMethodIdRef(this)
    }

    override fun accept(visitor: PsiElementVisitor) {
        if (visitor is GdVisitor) accept(visitor)
        else super.accept(visitor)
    }

    override fun getOwnReferences(): @Unmodifiable Collection<PsiSymbolReference> =
        polySymbolOwnReferences(this) {
            val requireStatic = PsiTreeUtil.getStubOrPsiParentOfType(
                this@GdGetMethodIdRefImpl,
                GdClassVarDeclTl::class.java
            )?.isStatic == true
            resolveFromNameMatchQuery(METHOD, text) { symbol ->
                !requireStatic || symbol.hasModifier(STATIC)
            }
        }

}
